src/lib/services/client-metrics/index.js

```diff
@@ -75,30 +78,56 @@ module.exports = class ClientMetricsService {
         });
     }
 
-    async upsertApp(value, clientIp) {
-        try {
-            const app = await this.clientAppStore.getApplication(value.appName);
-            await this.updateRow(value, app);
-        } catch (error) {
-            if (error instanceof NotFoundError) {
-                await this.clientAppStore.insertNewRow(value);
-                await this.eventStore.store({
-                    type: APPLICATION_CREATED,
-                    createdBy: clientIp,
-                    data: value,
-                });
+    async registerClient(data, clientIp) {
+        const value = await clientRegisterSchema.validateAsync(data);
+        value.clientIp = clientIp;
+        this.logger.info(`${JSON.stringify(data)}`);
+        this.seenClients[this.clientKey(value)] = value;
+    }
+
+    clientKey(client) {
+        return `${client.appName}_${client.instanceId}`;
+    }
+
+    async bulkAdd() {
+        if (
+            this &&
+            this.seenClients &&
+            this.clientAppStore &&
+            this.clientInstanceStore
+        ) {
+            const uniqueRegistrations = Object.values(this.seenClients);
+            const uniqueApps = Object.values(
+                uniqueRegistrations.reduce((soFar, reg) => {
+                    soFar[reg.appName] = reg;
+                    return soFar;
+                }, {}),
+            );
+            this.seenClients = {};
```

Comment:
Should seenClients only be emptied if the try succeeds? Eg. after the bulkUpserts? If they fail for some reason we'd have emptied the cache without inserting.

Reply:
The problem with that is what happens with the ones that are inserted into seenClients while we're doing the bulkUpsert? We'll zero the seenClients and might've added new ones we never saw. 
If we absolutely have to prevent losing these metrics, I'd say a better way is retrying on the same list we've copied out from the cache. 
Trying once more on error should be fine. Even if we inserted some, we're doing upserts with the same data, which should be fine

Reply:
Makes sense, thanks for the clarification.